agent/enhanced_tool_calling.py

```
@@ -0,0 +1,582 @@
+
+        report = self.feedback_generator.get_efficiency_report()
+
+        # Analyze tool usage patterns
+        for tool_name, stats in report.get("by_tool", {}).items():
+            # High redundancy tools
+            if stats["total"] > 0 and stats["redundant"] / stats["total"] > 0.3:
+                suggestions["tool_description_updates"].append({
+                    "tool": tool_name,
+                    "issue": "High redundancy rate",
+                    "suggestion": (
+                        f"Add to 'avoid_when': 'Result from previous call can be reused'"
```

Flake8 (7.3.0)

[error] f-string is missing placeholders

(F541)